client/relay-chain-rpc-interface/src/rpc_client.rs

```diff
@@ -0,0 +1,454 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+use backoff::{future::retry_notify, ExponentialBackoff};
+use cumulus_primitives_core::{
+	relay_chain::{
+		v2::{CommittedCandidateReceipt, OccupiedCoreAssumption, SessionIndex, ValidatorId},
+		Hash as PHash, Header as PHeader, InboundHrmpMessage,
+	},
+	InboundDownwardMessage, ParaId, PersistedValidationData,
+};
+use cumulus_relay_chain_interface::{RelayChainError, RelayChainResult};
+use futures::{
+	channel::mpsc::{Receiver, Sender},
+	StreamExt,
+};
+use jsonrpsee::{
+	core::{
+		client::{Client as JsonRPCClient, ClientT, Subscription, SubscriptionClientT},
+		Error as JsonRpseeError,
+	},
+	rpc_params,
+	types::ParamsSer,
+	ws_client::WsClientBuilder,
+};
+use parity_scale_codec::{Decode, Encode};
+use sc_client_api::StorageData;
+use sc_rpc_api::{state::ReadProof, system::Health};
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+use sp_core::sp_std::collections::btree_map::BTreeMap;
+use sp_runtime::DeserializeOwned;
+use sp_storage::StorageKey;
+use std::sync::Arc;
+
+pub use url::Url;
+
+const LOG_TARGET: &str = "relay-chain-rpc-client";
+
+const NOTIFICATION_CHANNEL_SIZE_LIMIT: usize = 20;
+
+/// Client that maps RPC methods and deserializes results
+#[derive(Clone)]
+pub struct RelayChainRPCClient {
+	/// Websocket client to make calls
+	ws_client: Arc<JsonRPCClient>,
+
+	/// Retry strategy that should be used for requests and subscriptions
+	retry_strategy: ExponentialBackoff,
+
+	to_worker_channel: Option<TracingUnboundedSender<NotificationRegisterMessage>>,
+}
+
+/// Worker messages to register new notification listeners
+#[derive(Clone, Debug)]
+pub enum NotificationRegisterMessage {
+	RegisterBestHeadListener(Sender<PHeader>),
+	RegisterImportListener(Sender<PHeader>),
+	RegisterFinalizationListener(Sender<PHeader>),
+}
+
+/// Worker that should be used in combination with [`RelayChainRPCClient`]. Must be polled to distribute header notifications to listeners.
+pub struct RPCStreamWorker {
+	// Communication channel with the RPC client
+	client_receiver: TracingUnboundedReceiver<NotificationRegisterMessage>,
+
+	// Senders to distribute incoming header notifications to
+	imported_header_listeners: Vec<Sender<PHeader>>,
+	finalized_header_listeners: Vec<Sender<PHeader>>,
+	best_header_listeners: Vec<Sender<PHeader>>,
+
+	// Incoming notification subscriptions
+	rpc_imported_header_subscription: Subscription<PHeader>,
+	rpc_finalized_header_subscription: Subscription<PHeader>,
+	rpc_best_header_subscription: Subscription<PHeader>,
+}
+
+/// Entry point to create [`RelayChainRPCClient`] and [`RPCStreamWorker`];
+pub async fn create_worker_client(
```

Comment:
Yes, it is expected that users of this method spawn the worker task immediately. Otherwise the streams will not work and all usages inside the collator need the streams to work. I will look into spawning directly.

As for the users of this, there will be other usage soon. Currently we immediately wrap the client in a `RelayChainRPCInterface`. In the future we will pass the client into the relay chain subsystems additionally.